accumulate.hpp

```diff
@@ -52,14 +52,15 @@ class iter::impl::Accumulator {
           sub_end_{std::move(sub_end)},
           accumulate_func_(&accumulate_fun),
           // only get first value if not an end iterator
-          acc_val_{
-              !(sub_iter_ != sub_end_) ? nullptr : new AccumVal(*sub_iter_)} {}
+          acc_val_{!(sub_iter_ != sub_end_)
+                       ? std::nullopt
+                       : std::make_optional<AccumVal>(*sub_iter_)} {}
 
     Iterator(const Iterator& other)
         : sub_iter_{other.sub_iter_},
           sub_end_{other.sub_end_},
           accumulate_func_{other.accumulate_func_},
-          acc_val_{other.acc_val_ ? new AccumVal(*other.acc_val_) : nullptr} {}
+          acc_val_{other.acc_val_} {}
```

Comment:
I am just not certain if default copy assignment does the check
```
      if (this == &other) {
        return *this;
      }
```
But I guess even if it doesn't it is not very important.